File: backend/app/schemas/stock.py
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field


MarketType = Literal["A股", "港股", "创业板", "科创板", "美股"]
RecommendationType = Literal["buy", "watch", "hold_cautious", "avoid"]
RiskLevel = Literal["low", "medium", "high"]
TradeReviewAction = Literal["buy", "sell", "add", "reduce", "observe"]
FollowUpStatus = Literal["open", "in_progress", "closed"]


class StockItem(BaseModel):
    symbol: str
    name: str
    market: MarketType
    board: str
    exchange: str
    industry: str
    sector: str
    concepts: List[str]
    tags: List[str]
    price: float
    change_pct: float
    analyzed: bool
    score: int
    recommendation: RecommendationType
    market_cap: Optional[float] = None
    pe: Optional[float] = None
    net_profit: Optional[float] = None
    revenue: Optional[float] = None
    revenue_growth: Optional[float] = None
    revenue_growth_qoq: Optional[float] = None
    profit_growth: Optional[float] = None
    profit_growth_qoq: Optional[float] = None
    gross_margin: Optional[float] = None
    net_margin: Optional[float] = None
    roe: Optional[float] = None
    debt_ratio: Optional[float] = None
    is_st: bool = False
    has_dividend: bool = False
    dividend_years: int = 0
    latest_dividend_year: Optional[str] = None
    dividend_yield: Optional[float] = None
    ex_dividend_date: Optional[str] = None
    is_high_dividend: bool = False
    is_ex_dividend_soon: bool = False
    updated_at: Optional[str] = None


class FinancialReport(BaseModel):
    year: str
    revenue: float
    net_profit: float
    gross_margin: float
    net_margin: float
    roe: float
    debt_ratio: float
    operating_cashflow: float
    eps: float
    dividend_yield: float
    report_url: str


class ValuationHistoryPoint(BaseModel):
    year: str
    pe: float
    pb: float
    market_cap: float


class DividendRecord(BaseModel):
    year: str
    cash_dividend_per_share: float
    payout_ratio: float
    ex_dividend_date: str


class ShareholderRecord(BaseModel):
    name: str
    holder_type: str
    holding_ratio: float
    change_yoy: float


class PeerCompany(BaseModel):
    symbol: str
    name: str
    market: str
    pe: float
    revenue_growth: float
    roe: float
    market_cap: float
    comparison_view: str


class StockDataQuality(BaseModel):
    source: str
    price_source: str
    fundamentals_source: str
    coverage_score: float
    reliability_score: int
    is_enriched: bool
    updated_at: Optional[str] = None
    freshness_days: Optional[int] = None
    warnings: List[str]


class StockDetail(BaseModel):
    symbol: str
    name: str
    market: MarketType
    board: str
    exchange: str
    sector: str
    price: float
    change_pct: float
    analyzed: bool
    pe: float
    pb: float
    roe: float
    debt_ratio: float
    revenue_growth: float
    profit_growth: float
    momentum: float
    volatility: float
    news_sentiment: float
    currency: str
    market_cap: float
    free_float_cap: float
    turnover_rate: float
    amplitude: float
    avg_volume_5d: float
    avg_volume_20d: float
    support_price: float
    resistance_price: float
    company_full_name: str
    english_name: str
    listing_date: str
    company_website: str
    investor_relations_url: str
    exchange_profile_url: str
    quote_url: str
    headquarters: str
    legal_representative: str
    employees: int
    main_business: str
    products_services: List[str]
    industry_positioning: str
    business_scope: List[str]
    market_coverage: List[str]
    company_intro: str
    business_tags: List[str]
    company_highlights: List[str]
    recent_events: List[str]
    core_logic: List[str]
    key_risks: List[str]
    catalyst_events: List[str]
    financial_reports: List[FinancialReport]
    valuation_history: List[ValuationHistoryPoint]
    dividend_history: List[DividendRecord]
    shareholder_structure: List[ShareholderRecord]
    peer_companies: List[PeerCompany]
    news_highlights: List[str]
    last_report_date: str
    data_quality: StockDataQuality


class TradePlan(BaseModel):
    entry_range: str
    stop_loss: str
    take_profit: str
    position_advice: str


class StockFactorScores(BaseModel):
    fundamental: int
    valuation: int
    momentum: int
    sentiment: int
    risk_control: int


class StockAnalysis(BaseModel):
    symbol: str
    score: int
    risk_level: RiskLevel
    recommendation: RecommendationType
    summary: str
    confidence: int
    methodology: str
    evidence_points: List[str]
    suitability_note: str
    disclaimer: str
    factor_scores: StockFactorScores
    strengths: List[str]
    risks: List[str]
    action_items: List[str]
    monitoring_points: List[str]
    scenario_analysis: List[str]
    trade_plan: TradePlan


class StockSnapshot(BaseModel):
    symbol: str
    detail: StockDetail
    analysis: StockAnalysis


class StockQAMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str = Field(min_length=1, max_length=2000)


class StockQARequest(BaseModel):
    question: str = Field(min_length=1, max_length=2000)
    history: List[StockQAMessage] = Field(default_factory=list, max_length=12)


class StockQAResponse(BaseModel):
    symbol: str
    question: str
    answer: str
    confidence: int = Field(ge=1, le=99)
    bullets: List[str]
    references: List[str]
    follow_up_questions: List[str]
    disclaimer: str
    generated_at: str
    search_used: bool = False
    search_query: Optional[str] = None
    search_result_count: int = 0


class StockTradeReviewBase(BaseModel):
    trade_date: str
    action: TradeReviewAction
    price: Optional[float] = None
    quantity: Optional[float] = None
    thesis: str
    execution_notes: Optional[str] = None
    outcome_review: Optional[str] = None
    lessons_learned: Optional[str] = None
    follow_up_items: List[str] = Field(default_factory=list)
    follow_up_status: FollowUpStatus = "open"
    next_review_date: Optional[str] = None
    confidence_score: Optional[int] = Field(default=None, ge=0, le=100)
    discipline_score: Optional[int] = Field(default=None, ge=0, le=100)


class StockTradeReviewCreate(StockTradeReviewBase):
    pass


class StockTradeReviewUpdate(BaseModel):
    trade_date: Optional[str] = None
    action: Optional[TradeReviewAction] = None
    price: Optional[float] = None
    quantity: Optional[float] = None
    thesis: Optional[str] = None
    execution_notes: Optional[str] = None
    outcome_review: Optional[str] = None
    lessons_learned: Optional[str] = None
    follow_up_items: Optional[List[str]] = None
    follow_up_status: Optional[FollowUpStatus] = None
    next_review_date: Optional[str] = None
    confidence_score: Optional[int] = Field(default=None, ge=0, le=100)
    discipline_score: Optional[int] = Field(default=None, ge=0, le=100)


class StockTradeReviewItem(StockTradeReviewBase):
    id: int
    symbol: str
    owner_user_id: Optional[int] = None
    owner_username: Optional[str] = None
    created_at: str
    updated_at: str
    floating_pnl: Optional[float] = None
    floating_pnl_pct: Optional[float] = None
    is_follow_up_due: bool = False


class StockTradeReviewSummary(BaseModel):
    total_reviews: int
    open_follow_ups: int
    in_progress_follow_ups: int
    closed_follow_ups: int
    due_follow_ups: int
    avg_confidence_score: float
    avg_discipline_score: float
    net_floating_pnl: float
    net_floating_pnl_pct: float


class StockTradeReviewResponse(BaseModel):
    symbol: str
    current_price: Optional[float] = None
    items: List[StockTradeReviewItem]
    summary: StockTradeReviewSummary


class DashboardSummary(BaseModel):
    total_stocks: int
    analyzed_count: int
    risk_alert_count: int
    average_score: float
    best_opportunities: List[StockItem]
    high_risk_stocks: List[StockItem]
    latest_updates: List[str]


class StockListStats(BaseModel):
    average_score: float
    median_score: float
    positive_change_count: int
    negative_change_count: int
    by_market: Dict[str, int]
    by_board: Dict[str, int]
    by_recommendation: Dict[str, int]
    top_industries: Dict[str, int]
    top_concepts: Dict[str, int]
    top_tags: Dict[str, int]


class StockDividendSummary(BaseModel):
    total: int
    continuous_3y_count: int
    high_yield_count: int
    upcoming_ex_dividend_count: int
    latest_year: Optional[str] = None
    by_market: Dict[str, int]
    by_board: Dict[str, int]


class StockListResponse(BaseModel):
    items: List[StockItem]
    total: int
    page: int
    page_size: int
    industries: Optional[List[str]] = None
    concepts: Optional[List[str]] = None
    tags: Optional[List[str]] = None
    boards: Optional[List[str]] = None
    exchanges: Optional[List[str]] = None
    recommendations: Optional[List[RecommendationType]] = None
    stats: Optional[StockListStats] = None
    dividend_summary: Optional[StockDividendSummary] = None
    last_synced_at: Optional[str] = None


class SectorConceptItem(BaseModel):
    name: str
    stock_count: int
    avg_change_pct: float
    relative_change_pct: float
    avg_score: float
    buy_watch_ratio: float
    breadth_ratio: float
    leader_avg_score: float
    heat_score: float
    confidence: int
    is_broad_theme: bool = False
    leading_symbols: List[str]
    rotation_stage: str
    warnings: List[str]


class SectorRotationResponse(BaseModel):
    generated_at: str
    market_scope: str
    benchmark_change_pct: float
    methodology: List[str]
    sample_policy: str
    total_sectors: int
    current_hot_sectors: List[SectorConceptItem]
    next_potential_sector: Optional[SectorConceptItem] = None
    rotation_path: List[str]
    reasoning: List[str]
    risk_warnings: List[str]


class KLinePoint(BaseModel):
    date: str
    open: float
    high: float
    low: float
    close: float
    volume: float


class StockKLineResponse(BaseModel):
    symbol: str
    period: str
    interval: str
    source: str
    points: List[KLinePoint]
    latest_price: Optional[float] = None
    change_pct: Optional[float] = None
    is_fallback: bool = False
    warning: Optional[str] = None


class StockSyncResponse(BaseModel):
    success: bool
    total_count: int
    a_share_count: int
    hk_count: int
    us_count: int = 0
    duration_ms: int
    last_synced_at: Optional[str] = None
    message: Optional[str] = None


class StockEnrichResponse(BaseModel):
    success: bool
    total: int
    processed: int
    success_count: int
    failed_count: int
    skipped_count: int
    duration_ms: int
    message: Optional[str] = None


class StockEnrichStatusResponse(BaseModel):
    total_universe: int
    enriched_count: int
    success_count: int
    failed_count: int
    coverage_rate: float
    latest_enriched_at: Optional[str] = None


class ReportListItem(BaseModel):
    report_id: str
    symbol: str
    name: str
    market: MarketType
    score: int
    recommendation: RecommendationType
    risk_level: RiskLevel
    report_version: str
    report_date: str
    headline: str


class ReportDetail(ReportListItem):
    summary: str
    key_points: List[str]
    risk_alerts: List[str]
    action_plan: List[str]
    trade_plan: TradePlan


class StockQuery(BaseModel):
    analyzed: Optional[bool] = None
    market: Optional[MarketType] = None
    board: Optional[str] = None
    exchange: Optional[str] = None
    industry: Optional[str] = None
    concept: Optional[str] = None
    tag: Optional[str] = None
    recommendation: Optional[RecommendationType] = None
    market_cap_min: Optional[float] = None
    market_cap_max: Optional[float] = None
    pe_max: Optional[float] = None
    price_min: Optional[float] = None
    price_max: Optional[float] = None
    net_profit_min: Optional[float] = None
    revenue_min: Optional[float] = None
    revenue_growth_min: Optional[float] = None
    revenue_growth_qoq_min: Optional[float] = None
    profit_growth_min: Optional[float] = None
    profit_growth_qoq_min: Optional[float] = None
    gross_margin_min: Optional[float] = None
    net_margin_min: Optional[float] = None
    exclude_st: Optional[bool] = None
    roe_min: Optional[float] = None
    debt_ratio_max: Optional[float] = None
    score_min: Optional[int] = None
    score_max: Optional[int] = None
    change_pct_min: Optional[float] = None
    change_pct_max: Optional[float] = None
    q: Optional[str] = None
    sort_by: Optional[Literal["score", "change_pct", "price"]] = "score"

File: frontend/src/types/stock.ts
export type MarketType = "A股" | "港股" | "创业板" | "科创板" | "美股";
export type RecommendationType = "buy" | "watch" | "hold_cautious" | "avoid";
export type RiskLevel = "low" | "medium" | "high";
export type MainForceStage = "accumulation" | "pullback" | "markup" | "distribution" | "neutral";
export type MainForceSignalLevel = "high" | "medium" | "low";
export type TradeReviewAction = "buy" | "sell" | "add" | "reduce" | "observe";
export type FollowUpStatus = "open" | "in_progress" | "closed";

export interface StockItem {
  symbol: string;
  name: string;
  market: MarketType;
  board: string;
  exchange: string;
  industry: string;
  sector: string;
  concepts: string[];
  tags: string[];
  price: number;
  change_pct: number;
  analyzed: boolean;
  score: number;
  recommendation: RecommendationType;
  market_cap?: number | null;
  pe?: number | null;
  net_profit?: number | null;
  revenue?: number | null;
  revenue_growth?: number | null;
  revenue_growth_qoq?: number | null;
  profit_growth?: number | null;
  profit_growth_qoq?: number | null;
  gross_margin?: number | null;
  net_margin?: number | null;
  roe?: number | null;
  debt_ratio?: number | null;
  is_st?: boolean;
  has_dividend?: boolean;
  dividend_years?: number;
  latest_dividend_year?: string | null;
  dividend_yield?: number | null;
  ex_dividend_date?: string | null;
  is_high_dividend?: boolean;
  is_ex_dividend_soon?: boolean;
  updated_at?: string | null;
}

export interface MainForceMetrics {
  range_20: number;
  range_60: number;
  range_squeeze: number;
  vol_20: number;
  vol_60: number;
  vol_squeeze: number;
  obv_slope_20: number;
  up_volume_ratio_20: number;
  close_slope_60: number;
  drawdown_60: number;
}

export interface MainForceCandidate {
  symbol: string;
  name: string;
  market: MarketType;
  price: number;
  change_pct: number;
  score: number;
  stage: MainForceStage;
  signal_level: MainForceSignalLevel;
  reason: string;
  metrics: MainForceMetrics;
  sentiment_score?: number | null;
  sentiment_summary?: string | null;
  sentiment_sources?: number;
  llm_summary?: string | null;
}

export interface MainForceScanResponse {
  generated_at: string;
  total_scanned: number;
  candidates: MainForceCandidate[];
}

export interface MainForceSettingResponse {
  enabled: boolean;
  scan_interval_minutes: number;
  overrides: Record<string, number | boolean>;
  effective: Record<string, number | boolean>;
  last_run_at?: string | null;
}

export interface MainForceSettingUpdate {
  enabled?: boolean;
  scan_interval_minutes?: number;
  overrides?: Record<string, number | boolean | null>;
}

export interface FinancialReport {
  year: string;
  revenue: number;
  net_profit: number;
  gross_margin: number;
  net_margin: number;
  roe: number;
  debt_ratio: number;
  operating_cashflow: number;
  eps: number;
  dividend_yield: number;
  report_url: string;
}

export interface ValuationHistoryPoint {
  year: string;
  pe: number;
  pb: number;
  market_cap: number;
}

export interface DividendRecord {
  year: string;
  cash_dividend_per_share: number;
  payout_ratio: number;
  ex_dividend_date: string;
}

export interface ShareholderRecord {
  name: string;
  holder_type: string;
  holding_ratio: number;
  change_yoy: number;
}

export interface PeerCompany {
  symbol: string;
  name: string;
  market: string;
  pe: number;
  revenue_growth: number;
  roe: number;
  market_cap: number;
  comparison_view: string;
}

export interface StockDataQuality {
  source: string;
  price_source: string;
  fundamentals_source: string;
  coverage_score: number;
  reliability_score: number;
  is_enriched: boolean;
  updated_at?: string | null;
  freshness_days?: number | null;
  warnings: string[];
}

export interface StockDetail {
  symbol: string;
  name: string;
  market: MarketType;
  board: string;
  exchange: string;
  sector: string;
  price: number;
  change_pct: number;
  analyzed: boolean;
  pe: number;
  pb: number;
  roe: number;
  debt_ratio: number;
  revenue_growth: number;
  profit_growth: number;
  momentum: number;
  volatility: number;
  news_sentiment: number;
  currency: string;
  market_cap: number;
  free_float_cap: number;
  turnover_rate: number;
  amplitude: number;
  avg_volume_5d: number;
  avg_volume_20d: number;
  support_price: number;
  resistance_price: number;
  company_full_name: string;
  english_name: string;
  listing_date: string;
  company_website: string;
  investor_relations_url: string;
  exchange_profile_url: string;
  quote_url: string;
  headquarters: string;
  legal_representative: string;
  employees: number;
  main_business: string;
  products_services: string[];
  industry_positioning: string;
  business_scope: string[];
  market_coverage: string[];
  company_intro: string;
  business_tags: string[];
  company_highlights: string[];
  recent_events: string[];
  core_logic: string[];
  key_risks: string[];
  catalyst_events: string[];
  financial_reports: FinancialReport[];
  valuation_history: ValuationHistoryPoint[];
  dividend_history: DividendRecord[];
  shareholder_structure: ShareholderRecord[];
  peer_companies: PeerCompany[];
  news_highlights: string[];
  last_report_date: string;
  data_quality: StockDataQuality;
}

export interface TradePlan {
  entry_range: string;
  stop_loss: string;
  take_profit: string;
  position_advice: string;
}

export interface StockFactorScores {
  fundamental: number;
  valuation: number;
  momentum: number;
  sentiment: number;
  risk_control: number;
}

export interface StockAnalysis {
  symbol: string;
  score: number;
  risk_level: RiskLevel;
  recommendation: RecommendationType;
  summary: string;
  confidence: number;
  methodology: string;
  evidence_points: string[];
  suitability_note: string;
  disclaimer: string;
  factor_scores: StockFactorScores;
  strengths: string[];
  risks: string[];
  action_items: string[];
  monitoring_points: string[];
  scenario_analysis: string[];
  trade_plan: TradePlan;
}

export interface StockSnapshot {
  symbol: string;
  detail: StockDetail;
  analysis: StockAnalysis;
}

export type StockQARole = "user" | "assistant";

export interface StockQAMessage {
  role: StockQARole;
  content: string;
}

export interface StockQARequest {
  question: string;
  history?: StockQAMessage[];
}

export interface StockQAResponse {
  symbol: string;
  question: string;
  answer: string;
  confidence: number;
  bullets: string[];
  references: string[];
  follow_up_questions: string[];
  disclaimer: string;
  generated_at: string;
  search_used?: boolean;
  search_query?: string | null;
  search_result_count?: number;
}

export interface KLinePoint {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface StockKLineResponse {
  symbol: string;
  period: "1mo" | "3mo" | "6mo" | "1y" | "5y";
  interval: "1d" | "1h";
  source: string;
  points: KLinePoint[];
  latest_price?: number | null;
  change_pct?: number | null;
  is_fallback: boolean;
  warning?: string | null;
}

export interface StockTradeReviewItem {
  id: number;
  symbol: string;
  owner_user_id?: number | null;
  owner_username?: string | null;
  trade_date: string;
  action: TradeReviewAction;
  price?: number | null;
  quantity?: number | null;
  thesis: string;
  execution_notes?: string | null;
  outcome_review?: string | null;
  lessons_learned?: string | null;
  follow_up_items: string[];
  follow_up_status: FollowUpStatus;
  next_review_date?: string | null;
  confidence_score?: number | null;
  discipline_score?: number | null;
  created_at: string;
  updated_at: string;
  floating_pnl?: number | null;
  floating_pnl_pct?: number | null;
  is_follow_up_due: boolean;
}

export interface StockTradeReviewSummary {
  total_reviews: number;
  open_follow_ups: number;
  in_progress_follow_ups: number;
  closed_follow_ups: number;
  due_follow_ups: number;
  avg_confidence_score: number;
  avg_discipline_score: number;
  net_floating_pnl: number;
  net_floating_pnl_pct: number;
}

export interface StockTradeReviewResponse {
  symbol: string;
  current_price?: number | null;
  items: StockTradeReviewItem[];
  summary: StockTradeReviewSummary;
}

export interface StockTradeReviewCreateRequest {
  trade_date: string;
  action: TradeReviewAction;
  price?: number;
  quantity?: number;
  thesis: string;
  execution_notes?: string;
  outcome_review?: string;
  lessons_learned?: string;
  follow_up_items?: string[];
  follow_up_status?: FollowUpStatus;
  next_review_date?: string;
  confidence_score?: number;
  discipline_score?: number;
}

export interface StockTradeReviewUpdateRequest {
  trade_date?: string;
  action?: TradeReviewAction;
  price?: number;
  quantity?: number;
  thesis?: string;
  execution_notes?: string;
  outcome_review?: string;
  lessons_learned?: string;
  follow_up_items?: string[];
  follow_up_status?: FollowUpStatus;
  next_review_date?: string;
  confidence_score?: number;
  discipline_score?: number;
}

export interface DashboardSummary {
  total_stocks: number;
  analyzed_count: number;
  risk_alert_count: number;
  average_score: number;
  best_opportunities: StockItem[];
  high_risk_stocks: StockItem[];
  latest_updates: string[];
}

export interface StockListStats {
  average_score: number;
  median_score: number;
  positive_change_count: number;
  negative_change_count: number;
  by_market: Record<string, number>;
  by_board: Record<string, number>;
  by_recommendation: Record<string, number>;
  top_industries: Record<string, number>;
  top_concepts: Record<string, number>;
  top_tags: Record<string, number>;
}

export interface StockDividendSummary {
  total: number;
  continuous_3y_count: number;
  high_yield_count: number;
  upcoming_ex_dividend_count: number;
  latest_year?: string | null;
  by_market: Record<string, number>;
  by_board: Record<string, number>;
}

export interface StockListResponse {
  items: StockItem[];
  total: number;
  page: number;
  page_size: number;
  industries?: string[] | null;
  concepts?: string[] | null;
  tags?: string[] | null;
  boards?: string[] | null;
  exchanges?: string[] | null;
  recommendations?: RecommendationType[] | null;
  stats?: StockListStats | null;
  dividend_summary?: StockDividendSummary | null;
  last_synced_at?: string | null;
}

export interface SectorConceptItem {
  name: string;
  stock_count: number;
  avg_change_pct: number;
  relative_change_pct: number;
  avg_score: number;
  buy_watch_ratio: number;
  breadth_ratio: number;
  leader_avg_score: number;
  heat_score: number;
  confidence: number;
  is_broad_theme: boolean;
  leading_symbols: string[];
  rotation_stage: string;
  warnings: string[];
}

export interface SectorRotationResponse {
  generated_at: string;
  market_scope: string;
  benchmark_change_pct: number;
  methodology: string[];
  sample_policy: string;
  total_sectors: number;
  current_hot_sectors: SectorConceptItem[];
  next_potential_sector?: SectorConceptItem | null;
  rotation_path: string[];
  reasoning: string[];
  risk_warnings: string[];
}

export interface StockSyncResponse {
  success: boolean;
  total_count: number;
  a_share_count: number;
  hk_count: number;
  us_count: number;
  duration_ms: number;
  last_synced_at?: string | null;
  message?: string | null;
}

export interface StockEnrichResponse {
  success: boolean;
  total: number;
  processed: number;
  success_count: number;
  failed_count: number;
  skipped_count: number;
  duration_ms: number;
  message?: string | null;
}

export interface StockEnrichStatusResponse {
  total_universe: number;
  enriched_count: number;
  success_count: number;
  failed_count: number;
  coverage_rate: number;
  latest_enriched_at?: string | null;
}

export interface ReportListItem {
  report_id: string;
  symbol: string;
  name: string;
  market: MarketType;
  score: number;
  recommendation: RecommendationType;
  risk_level: RiskLevel;
  report_version: string;
  report_date: string;
  headline: string;
}

export interface ReportDetail extends ReportListItem {
  summary: string;
  key_points: string[];
  risk_alerts: string[];
  action_plan: string[];
  trade_plan: TradePlan;
}
